Highlighting of columns

paretje opened the pull request:
This PR implements #212. If the entry has a column assigned, it is highlighted, if no column is specified, nothing is done, as this would just duplicate information already visualized by the signs.

Writing this PR, I see however a comment in the issue to visualise lines. Maybe I should make it an option to highlight the complete line when no column is known. Then you could use this instead of the signs. It's easy to do, but I'll probably wait for a comment before I do.

blueyed commented:
@paretje 
Thanks for this!
Can you please rebase this on master (resolving conflics) and then also squash it?

paretje commented:
@blueyed: done

blueyed commented on autoload/neomake/highlights.vim:
```diff
@@ -0,0 +1,66 @@
+    call s:ResetHighlights()
+    for l:type in ['file', 'project']
+        for l:hi in keys(get(s:highlights[l:type], a:buf, {}))
+            call matchaddpos(l:hi, s:highlights[l:type][a:buf][l:hi])
```
This seems to return the id, that can be used with `matchdelete` - so there's no need to loop over all matches in `s:ResetHighlights`.

blueyed commented:
Some feedback:
1. errors are highlighted in the location list window for me, too!  (if there's an error in line 1, col 8, it will be highlighted also in the location list window).  Can you confirm this?  Otherwise it might be related to my location list handling (e.g. auto-resizing it).
2. I think using something like `cterm=underline` for the highlighting groups would be less distracting.
3. It should be reset/hidden for the current line if that gets changed (via `TextChanged/TextChangedI` maybe)?  Because the position will usually get out of sync then.

blueyed commented:
There are also highlights in a new terminal buffer (via `:terminal` or fzf (which uses a terminal) sometimes).
I still have not looked at the code itself in more depth..

paretje commented:
Ah, well, highlights are global in vim. They are removed when processing a buffer by `neomake#ProcessCurrentWindow`. I guess it's never called in a terminal buffer. Can be fixed by adding a autocmd to do so.

It currently uses caches by buffer. I know neomake switched to window variables, I guess it could be calculated on the fly using them. What was the exact reason to use window variables, and not ie buffer variables?

blueyed commented:
> What was the exact reason to use window variables, and not ie buffer variables?

IIRC it was in the context of having the same buffer in multiple windows.

Do you plan to update this PR (needs to be rebased, too).

paretje commented:
I'll update and rebase the PR next week.

mhartington commented:
Not sure where @paretje is with the update, but I have a rebased version of this that I've been testing locally. Would you like me to submit a new PR and go from there?

paretje commented:
I'm sorry for the delay, but I rebased the PR, and fixed the issue with terminal buffers. Id's are now kept track of, as you suggested. I remember I did try it before, but I believe I used a script variable, which of course didn't work, as the highlights are per window.

teto commented:
> Maybe I should make it an option to highlight the complete line when no column is known. Then you could use this instead of the signs
I wish to see that :) !
I've checked out the PR but don't see any change (latex or python), I guess I should run it on a C++ project as clang does return a column number ?

bounceme commented:
here is a helper for match creation:
```
function! neomake#highlight#MatchAdd(hi,pos) abort
  if exists('*matchaddpos')
    return matchaddpos(a:hi,a:pos)
  endif
  return matchadd(a:hi,'\%' . a:pos[0] . 'l\%' . a:pos[1] . 'c')
endfunction
```

JelteF commented:
I just created my own fork with these changes. I have also added support to supply the optional length of key by the maker to highlight more than just the first letter. I implemented this in a basic way for the flake8 maker. https://github.com/JelteF/neomake/tree/highlight (I also rebased it on current master, which could be done automatically)
I've only tested it very shortly, but it seems to work quite well.

Also, could someone with Travis permissions restart the build for this pull request, because I think the current failure might not be correct.

blueyed commented:
For Neovim there is an API for buffer highlight!
See `:h api-highlights`:

> Nvim allows plugins to add position-based highlights to buffers. This is
similar to |matchaddpos()| but with some key differences. The added highlights
are associated with a buffer and adapts to line insertions and deletions,
similar to signs. It is also possible to manage a set of highlights as a group
and delete or replace all at once.

I think this should be used here.

paretje commented:
@blueyed: OK, to gather everything, I guess I need to do this:
- [ ] support for new neovim api
- [ ] support to highlight complete line
- [ ] cherry-pick @JelteF's changes to support length in matches

@bounceme: why? `matchaddpos` has been added in 7.4.330, and neomake already requires 7.4.503.

bounceme commented:
up to you

paretje commented:
When I added the option to highlight the complete line, I realised I should add documentation for the various options. I added this to the TODO list.

paretje commented:
@blueyed: I think this PR is ready now. What do you think?

paretje commented:
Rebased on current master.

Alok commented:
+1 to this

JelteF commented on autoload/neomake.vim:
```diff
@@ -885,8 +894,14 @@ function! neomake#CleanOldFileSignsAndErrors(...) abort
         if has_key(s:current_errors['file'], bufnr)
             unlet s:current_errors['file'][bufnr]
         endif
+<<<<<<< HEAD
```
You forgot to merge this

blueyed commented on autoload/neomake/makers/ft/python.vim:
```diff
@@ -83,6 +83,11 @@ function! neomake#makers#ft#python#Flake8EntryProcess(entry) abort
+    let l:token = matchstr(a:entry.text, "'.*'")
+    if strlen(l:token)
+        let a:entry.length = strlen(l:token) - 2 " subtract the quotes
+    endif
```
@JelteF @paretje 
What is this needed for?
Should be a separate PR probably.

paretje replied:
@paretje: that's an extension of @JelteF to have highlights of variable lengths when makers produce such information. What should be separated? The implementation for python, or the whole feature?

blueyed replied:
Oh, it was not clear to me (and still is not): then length will only be set if the text is in quotes?
How does entry.length get set in general?  (is it only overwritten here?)

JelteF replied:
Length will indeed only be set if text is in quotes (indicating a specific word). It's not full proof but it works quite well for flake8 errors. I've been using this branch for a while now and the only error for which it is wrong is the unused import one.

Length can be set optionally, because not all linters/makers will be able to support it. However, for ones that do it can be really useful. I'm mostly working on python projects with flake8 so I added support for that one, allowing possible additions for other makers to be done later.

paretje replied:
Well, by default the length is one column. Some makers contain information which can be used for this, e.g. to mark a non-existing function. The only maker currently setting `length` is `flake8`.

In this case, when you have a file:
```python
#!/usr/bin/python
nonexist
```

`flake8 --format=default test.py`:
```
test.py:2:1: F821 undefined name 'nonexist'
```

blueyed replied:
Makes sense, thanks for explaining!
@JelteF 
I have a local branch/stash to improve Python makers btw.. all my recent improvements are ways to get there.. the current one being to fix "serialize": I would like to run the "python" maker serialized, and when it bails out (syntax errors) skip the others.

blueyed replied:
Found some issue regarding this: https://github.com/neomake/neomake/issues/824.

blueyed commented:
Tests are failing on Neovim (https://travis-ci.org/neomake/neomake/jobs/175132001#L779), which might be an issue with the tests though.

I would really like to see new tests for this feature, too.

paretje commented:
I have encountered the error from Travis before. I investigated it, and for a moment I thought it was a bug in neovim, but apparently the bufnr passed isn't always a number. It's easy to convert it back to a number, but I'll investigate what's happening.

paretje commented:
I found and fixed the issue. @blueyed: would something like https://github.com/neomake/neomake/blob/master/tests/signs.vader do for tests?

blueyed commented:
@paretje 
Not sure about the tests, but that's one of the later ones, so it might be a good base.  Thanks for considering to write them!

JelteF commented:
Is there anything holding this up to be merged? (except for the obvious merge conflict at the moment) I've been using this as my main branch for quite a while now and have not noticed any issues.

paretje commented:
As far as I know, I just have the write the tests.

blueyed commented:
Oops.. ;)

I got the impression that @paretje is willing to write tests after the fact instead.
(came here via the [Neovim #7 Newsletter](https://neovim.io/news/2016/11/), which mentioned a feature being used here).

Alok commented:
=)
